### PR TITLE
optimize documentdb query based on service team feedback

### PR DESCRIPTION
**Pull request by yungez**
@ZhijunZhao @aovechki 

1. avoid query collection, instead of use constructed collection link
2. use readDocument instead of queryDocument for performance consideration

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -26,6 +26,12 @@
             <name>Zhijun Zhao</name>
             <email>zhijzhao@microsoft.com</email>
         </developer>
+        
+         <developer>
```
Comment (ZhijunZhao) [inline pom.xml]: reformat? it seems the tab/space is different.

Comment (yungez) [inline pom.xml]: it's reformatted already, all space.

Comment (yungez) [inline pom.xml]: it's reformatted already, all space

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java [hunk 2]
```diff
@@ -285,6 +286,10 @@ private String getDatabaseLink(String databaseName) {
     }
 
     private String getCollectionLink(String databaseName, String collectionName) {
```
Comment (ZhijunZhao) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: Now that database name is a data member, I prefer to remove this parameter.

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java [hunk 3]
```diff
@@ -43,71 +46,145 @@ public DocumentDbTemplate(DocumentDbFactory documentDbFactory,
...
-    public <T> void delete(T objectToRemove) {
-        throw new UnsupportedOperationException("not supported");
+    public String getCollectionName(Class<?> entityClass) {
+
+        return entityClass.getSimpleName();
     }
 
-    private Database getDb(String dbName) {
+    private Database createOrGetDatabase(String dbName) {
         final String query = "SELECT * FROM root r WHERE r.id='" + dbName + "'";
```
Comment (ZhijunZhao) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: I think below code section 

                if (resource instanceof Database) {
                    db = (Database) resource;
                }
                return db;

should be changed to 

                if (resource instanceof Database) {
                    return  (Database) resource;
                } else {
                     return null;
                }

Although origin code works, the logic is not correct.

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java (hunk 2 above)
Comment (yungez) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: interface should be general. like collectionName, although could get from entityClass, but still use have choice to customize it

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java (hunk 3 above)
Comment (yungez) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: nice catch, fixed

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java [hunk 4]
```diff
@@ -158,32 +234,27 @@ public DocumentCollection createCollection(String collectionName, RequestOptions
...
+        final List<DocumentCollection> collectionList = documentDbFactory.getDocumentClient()
+                .queryCollections(getDatabaseLink(dbName), query, null).getQueryIterable().toList();
 
-                final RequestOptions requestOptions = new RequestOptions();
-                requestOptions.setOfferThroughput(1000);
+        if (collectionList.size() > 0) {
+            return collectionList.get(0);
+        } else {
+            DocumentCollection collection = new DocumentCollection();
+            collection.setId(collectionName);
```
Comment (ZhijunZhao) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: These two lines are not used.

Comment (yungez) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: fixed

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java [hunk 5]
```diff
@@ -202,76 +273,6 @@ public void dropCollection(String collectionName) {
...
-
-        for (int i = 0; i < results.size(); i++) {
-            final T entity = dbConverter.convertFromDocument(results.get(i), entityClass);
-            entities.add(entity);
-        }
-
-        return entities;
-    }
-
     public DocumentDbFactory getDocumentDbFactory() {
```
Comment (ZhijunZhao) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: getDocumentDbFactory and getDocumentDbConverter are not referenced.

Comment (yungez) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: no, they're not referenced.

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java [hunk 6]
```diff
@@ -158,32 +234,27 @@ public DocumentCollection createCollection(String collectionName, RequestOptions
 
     }
 
-    private DocumentCollection getCollection(String dbName, String collectionName) {
+    private DocumentCollection createOrGetCollection(String dbName, String collectionName) {
```
Comment (ZhijunZhao) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: A few functions like this don't have the logging code.

**Comment by ZhijunZhao** [comment]
It might be good to add a new API to DocumentDbRepository, as suggested by service team. 

FindAllByProperty

File: spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java (hunk 6 above)
Comment (yungez) [inline spring-data-azure-documentdb/src/main/java/com/microsoft/azure/spring/data/documentdb/core/DocumentDbTemplate.java]: nice. add missed logger info to write operations

**Comment by yungez** [comment]
there're lots of feature could be supported next, e.g paging, findall, delete, customized find, count, batch operation together with findByProperty. timeline is not determined yet based on roadmap.

**Pull request merged**